docs/csharp/nullable-attributes.md

```diff
@@ -0,0 +1,301 @@
+
+```csharp
+[AllowNull]
+public string ScreenName
+{
+   get => screenName;
+   set => screenName = value ?? GenerateRandomScreenName();
+}
+private string screenName = GenerateRandomScreenName();
+```
+
+You may need to add a `using` directive for <xref:System.Diagnostics.CodeAnalysis> to use this and other attributes discussed in this article. The attribute is applied to the property, not the `set` accessor. The `AllowNull` attribute specifies *pre-conditions*, and only applies to inputs. The `get` accessor has a return value, but no input arguments. Therefore, the `AllowNull` attribute only applies to the `set` accessor.
+
+The preceding example demonstrates what to look for when adding the `AllowNull` attribute on an argument:
+
+1. The general contract for that variable is that it shouldn't be `null`, so you want a non-nullable reference type.
+1. There are scenarios for the input variable to be `null`, though they aren't the most common usage.
+
+Most often you'll need this attribute for properties, or `in` `out` and `ref` arguments. The `AllowNull` attribute is the best choice when a variable is typically non-null, but you need to allow `null` as a precondition.
+
+Contrast that with scenarios for using `DisallowNull`: You use this attribute to specify that an input variable of a nullable type shouldn't be `null`. Consider a property where `null` is the default value, but clients can only set it to a non-null value. Consider the following code:
+
+```csharp
+public string ReviewComment // Comments can be set, but not cleared.
+{
+    get => _comment;
+    set
+    {
+        if (value == null) throw new ArgumentNullException(nameof(value), "Cannot set to null");
+        _comment = value;
+    }
+}
+string _comment;
+```
+
+The preceding code is the best way to express your design that the `ReviewComment` could be `null`, but can't be set to `null`. Once this code is nullable aware, you can express this concept more clearly to callers using the <xref:System.Diagnostics.CodeAnalysis.DisallowNullAttribute?displayProperty=nameWithType>:
+
+```csharp
+[DisallowNull] 
+public string? ReviewComment // Comments can be added, but not removed.
+{
+    get => _comment;
+    set => _comment = value ?? throw new ArgumentNullException(nameof(value), "Cannot set to null");;
+    }
+}
+string? _comment;
+```
+
+In a nullable context, the `ReviewComment` `get` accessor could return the default value of `null`. The compiler ensures that it must be checked before access. Furthermore, it warns callers that, even though it could be `null`, callers shouldn't explicitly set it to `null`. The `DisallowNull` attribute also specifies a *pre-condition*, it does not affect the `get` accessor. You should choose to use the `DisallowNull` attribute when you observe these characteristics about:
```

Comment:
"compiler ensures that it must be checked" might be understood as (1) "compiler adds the check"  or (2) the code wouldn't compile if there is no check
```suggestion
In a nullable context, the `ReviewComment` `get` accessor could return the default value of `null`. The compiler warns that it must be checked before access. Furthermore, it warns callers that, even though it could be `null`, callers shouldn't explicitly set it to `null`. The `DisallowNull` attribute also specifies a *pre-condition*, it does not affect the `get` accessor. You should choose to use the `DisallowNull` attribute when you observe these characteristics about:
```